src/main/java/com/atlassian/db/replica/spi/jdbc/JdbcUrl.java

```diff
@@ -0,0 +1,56 @@
+package com.atlassian.db.replica.spi.jdbc;
+
+public class JdbcUrl {
+    private static final String PREFIX = "jdbc";
+
+    private final JdbcProtocol protocol;
+    private final String endpoint;
+    private final String database;
+
+    public JdbcUrl(JdbcProtocol protocol, String endpoint, String database) {
+        this.protocol = protocol;
+        this.endpoint = endpoint;
+        this.database = database;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public String toString() {
+        return String.format("%s:%s://%s/%s", PREFIX, protocol, endpoint, database);
+    }
+
+    public static final class Builder {
+        private JdbcProtocol protocol;
+        private String endpoint;
+        private String database;
+
+        private Builder() {
+        }
+
+        public Builder protocol(JdbcProtocol protocol) {
+            this.protocol = protocol;
+            return this;
+        }
+
+        public Builder endpoint(String endpoint) {
+            this.endpoint = endpoint;
+            return this;
+        }
+
+        public Builder database(String database) {
+            this.database = database;
+            return this;
+        }
+
+        public JdbcUrl build() {
+            return new JdbcUrl(protocol, endpoint, database);
+        }
+
+        public String toString() {
```

Comment:
do we need toString in the builder?

Reply:
Does it matter? 😉

Reply:
other builders in the repository do not override the `toString` method so I guess I could remove it for consistency's sake.
otoh I think it is a good practice to always provide the `toString` implementation as the default one is very poor

Reply:
Yes. The default implementation is useless. I think @tciesla-atlassian meant there's a high chance "You aren't gonna need it".

Reply:
true, will remove it then